File: mini-util/src/main/java/com/mini/core/util/tree/ITree.java
package com.mini.core.util.tree;

import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.List;

public interface ITree<T extends ITree<T>> {
    default void addChild(@NotNull T child) {
        if (this.getChildren() == null) {
            var c = new ArrayList<T>();
            setChildren(c);
        }
        getChildren().add(child);
    }

    void setChildren(List<T> children);

    List<T> getChildren();

    Long getParentId();

    String getName();

    Long getId();
}

File: mini-util/src/main/kotlin/com/mini/core/util/tree/ITree.kt
@file:Suppress("unused")

package com.mini.core.util.tree

interface ITree<T : ITree<T>> {
    val children: MutableList<T>

    fun addChild(child: T) {
        children.add(child)
    }

    var parentId: Long?
    var name: String?
    var id: Long
}